data/trie/node_test.go

```diff
@@ -640,3 +646,144 @@ func TestTrieAddHashesToOldHashes(t *testing.T) {
 	tr.AppendToOldHashes(hashes)
 	assert.Equal(t, expectedHLength, len(tr.oldHashes))
 }
+
+func TestRecreateTrieFromSnapshotDb(t *testing.T) {
+	t.Parallel()
+
+	testVals := []struct {
+		key   []byte
+		value []byte
+	}{
+		{[]byte("doe"), []byte("reindeer")},
+		{[]byte("dog"), []byte("puppy")},
+		{[]byte("dogglesworth"), []byte("cat")},
+	}
+
+	msh, hsh := getTestMarshAndHasher()
+
+	tempDir, _ := ioutil.TempDir("", "leveldb_temp")
+	cfg := config.DBConfig{
+		FilePath:          tempDir,
+		Type:              string(storageUnit.LvlDbSerial),
+		BatchDelaySeconds: 1,
+		MaxBatchSize:      1,
+		MaxOpenFiles:      10,
+	}
+
+	tr := &patriciaMerkleTrie{
+		db:            mock.NewMemDbMock(),
+		snapshots:     make([]data.DBWriteCacher, 0),
+		snapshotDbCfg: cfg,
+		marshalizer:   msh,
+		hasher:        hsh,
+	}
+
+	for _, testVal := range testVals {
+		_ = tr.Update(testVal.key, testVal.value)
+	}
+
+	_ = tr.Snapshot()
+	collapsedRoot, _ := tr.root.getCollapsed(tr.marshalizer, tr.hasher)
+
+	snapshotTrie := &patriciaMerkleTrie{
+		root:        collapsedRoot,
+		db:          tr.snapshots[0],
+		marshalizer: msh,
+		hasher:      hsh,
+	}
+
+	for _, testVal := range testVals {
+		val, err := snapshotTrie.Get(testVal.key)
+		assert.Nil(t, err)
+		assert.Equal(t, testVal.value, val)
+	}
+}
+
+func TestEachSnapshotCreatesOwnDatabase(t *testing.T) {
+	t.Parallel()
+
+	testVals := []struct {
+		key   []byte
+		value []byte
+	}{
+		{[]byte("doe"), []byte("reindeer")},
+		{[]byte("dog"), []byte("puppy")},
+		{[]byte("dogglesworth"), []byte("cat")},
+	}
+
+	msh, hsh := getTestMarshAndHasher()
+
+	tempDir, _ := ioutil.TempDir("", "leveldb_temp")
+	cfg := config.DBConfig{
+		FilePath:          tempDir,
+		Type:              string(storageUnit.LvlDbSerial),
+		BatchDelaySeconds: 1,
+		MaxBatchSize:      1,
+		MaxOpenFiles:      10,
+	}
+
+	tr := &patriciaMerkleTrie{
+		db:            mock.NewMemDbMock(),
+		snapshots:     make([]data.DBWriteCacher, 0),
+		snapshotId:    0,
+		snapshotDbCfg: cfg,
+		marshalizer:   msh,
+		hasher:        hsh,
+	}
+
+	for _, testVal := range testVals {
+		_ = tr.Update(testVal.key, testVal.value)
+		_ = tr.Snapshot()
+
+		snapshotId := strconv.Itoa(tr.snapshotId - 1)
```

Comment:
maybe `fmt.Sprintf()` instead of `strconv` ?

Reply:
Just benchmarked `fmt.Sprintf()` and `strconv`. `fmt.Sprintf()` is 3 times more slower than `strconv`.

Reply:
ok. good to know